src/main/java/com/google/devtools/build/lib/skyframe/ToolchainUtil.java

```diff
@@ -63,6 +65,7 @@ static ToolchainContext createToolchainContext(
       Environment env,
       String targetDescription,
       Set<Label> requiredToolchains,
+      Set<Label> execConstraintLabels,
```

Comment:
This function takes several pretty generic parameters, is static and quite long and complex. All this argues for making it its own class with a builder and more decomposition. That would also allow you to remove the parameter comments you had to use throughout the remainder of this change.